### PR TITLE
[dv] Always preload ROM in chip_base_vseq

### LINKED ISSUE #15893: [test-triage] chip_rv_dm_lc_disabled
### Hierarchy of regression failure

Chip Level

### Failure Description

```
UVM_FATAL @ * us: (csr_utils_pkg.sv:542) [csr_utils::csr_spinwait] timeout chip_reg_block.kmac.cfg_regwen (addr=*) == * has 2 failures:

Test chip_rv_dm_lc_disabled has 2 failures.
0.chip_rv_dm_lc_disabled.3567407281
Line 462, in log /container/opentitan-public/scratch/os_regression/chip_earlgrey_asic-sim-vcs/0.chip_rv_dm_lc_disabled/latest/run.log

  UVM_FATAL @ 18747.638964 us: (csr_utils_pkg.sv:542) [csr_utils::csr_spinwait] timeout chip_reg_block.kmac.cfg_regwen (addr=0x41120010) == 0x0
  UVM_INFO @ 18747.638964 us: (uvm_report_catcher.svh:705) [UVM/REPORT/CATCHER]
  --- UVM Report catcher Summary ---
  
  
2.chip_rv_dm_lc_disabled.2379868426
Line 438, in log /container/opentitan-public/scratch/os_regression/chip_earlgrey_asic-sim-vcs/2.chip_rv_dm_lc_disabled/latest/run.log

  UVM_FATAL @ 14852.578599 us: (csr_utils_pkg.sv:542) [csr_utils::csr_spinwait] timeout chip_reg_block.kmac.cfg_regwen (addr=0x41120010) == 0x0
  UVM_INFO @ 14852.578599 us: (uvm_report_catcher.svh:705) [UVM/REPORT/CATCHER]
  --- UVM Report catcher Summary ---

```

### Steps to Reproduce
- GitHub Revision: [53874025b](https://github.com/lowrisc/opentitan/tree/53874025b840788d5e66ffddae8fe7840a9cf4df)
- dvsim invocation command to reproduce the failure, inclusive of build and run seeds:
  ./util/dvsim/dvsim.py hw/top_earlgrey/dv/chip_sim_cfg.hjson -i chip_rv_dm_lc_disabled --fixed-seed 3567407281 --build-seed 137638316 --waves fsdb -v h

### Tests with similar or related failures

_No response_

### PR DESCRIPTION
**Pull request by msfschaffner**
#15794 introduced a ROM init done check to `post_apply_reset()` in `chip_stub_cpu_base_vseq` which caused the ROM init done check in `dut_init()` of the `chip_rv_dm_lc_disabled_vseq` to hang forever.

This PR removes the ROM init check from `chip_rv_dm_lc_disabled_vseq` to fix that. In addition, it always pre-loads a valid ROM image during the `dut_init()` phase of `chip_stub_cpu_base_vseq` to make sure that the ROM check can succeed successfully. Note that the real hardware will always contain a valid ROM image, so this is not a simulation-only hack and bring simulations closer to the behavior of the real hardware.

This fixes #15893.

Signed-off-by: Michael Schaffner <msf@google.com>

### CODE REVIEW

File: hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv
Comment (tjaychen) [inline hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:52]: curious question, if we are in disable integrity check, and try to load the 39b version onto the 32b rom, does it correctly truncate..?

Comment (tjaychen) [reply hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:52]: ah one last question, for the rom load, should we move this to `pre_start`? that way for tests that directly call reset, it won't end up trying to load on every reset. @sriyerg probably knows best where to put this.

File: hw/top_earlgrey/dv/chip_sim_cfg.hjson
Comment (tjaychen) [inline hw/top_earlgrey/dv/chip_sim_cfg.hjson:1424]: ah just to make sure i understand, since ROM is always loaded, we no longer need a `with_test_rom` option, is that right?

File: hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv
Comment (msfschaffner) [reply hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:52]: I believe so, yes. @matutem has added it as part of https://github.com/lowRISC/opentitan/pull/14594 to speed up local simulation iterations.

Comment (msfschaffner) [reply hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:52]: As for `dut_init()` - I did this together with @weicaiyang last week, and we replicated how the ROM loading has been handled in `chip_base_vseq` (all images are loaded in `dut_init()` in that sequence.

I am happy to change it though, if we think that would be better.

File: hw/top_earlgrey/dv/chip_sim_cfg.hjson
Comment (msfschaffner) [reply hw/top_earlgrey/dv/chip_sim_cfg.hjson:1424]: yes that is correct. I have added this special mode only for that test last week, but it probs makes sense to just use this for all stub CPU sequences so that we are simulating with an environment that resembles the "real" HW more closely.

Comment (tjaychen) [reply hw/top_earlgrey/dv/chip_sim_cfg.hjson:1424]: i have no idea if it matters at all in terms of simulation speed. hehe. Probably not important enough to really matter.
I would just merge as is, and if it is better for some reason we can tweak in the future.

**Review by tjaychen — APPROVED** [review]

File: hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv
Comment (weicaiyang) [reply hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:52]: sgtm, we can move this part to chip_base_vseq::pre_start and also delete the rom load in chip_sw_base_vseq::dut_init, so that we don't need to replicate it in 2 places.

**Comment by sriyerg** [comment]
If you dont mind, could you wait until my PR #15814 is merged?

File: hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv
Comment (sriyerg) [reply hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:52]: Agree with @weicaiyang - better to move this to `chip_base_vseq`.

File: hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv
Comment (sriyerg) [inline hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:53]: If ROM intg check is disabled, do we still need to preload something?

File: hw/top_earlgrey/dv/chip_sim_cfg.hjson
Comment (sriyerg) [inline hw/top_earlgrey/dv/chip_sim_cfg.hjson:257]: A future enhancement for M3 would be to preload the ROM with random data and compute and set the intg portion to the correct value, rather than preloading an actual test image. That would eliminate the extra bazel invocation step.

File: hw/top_earlgrey/dv/chip_sim_cfg.hjson
Comment (sriyerg) [inline hw/top_earlgrey/dv/chip_sim_cfg.hjson:254]: I think it will be good to expand on this and explain that the CPU is still, not going to fetch and execute anything in this mode since it is effectively disconnected.

**Comment by msfschaffner** [comment]
> If you dont mind, could you wait until my PR #15814 is merged?

It would be great if we could merge this fix so that the nightlies can pick this up... 
Do you expect yours to be merged soon?

File: hw/top_earlgrey/dv/chip_sim_cfg.hjson
Comment (sriyerg) [reply hw/top_earlgrey/dv/chip_sim_cfg.hjson:1424]: I provided a suggestion above - the stub CPU mode tests were able to start off immediately. But now, they are also going to be bottlenecked by the bazel step, which enforces serialization across all simulation instances. It is probably not a huge deal (maybe we can benchmark and see). But given the frequent bazel outages, it'd be great to minimize dependencies.

File: hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv
Comment (sriyerg) [inline hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:60]: [Thanks for moving this to `dut_init()`. ](https://github.com/lowRISC/opentitan/pull/15794/files#r1007477331)

File: hw/top_earlgrey/dv/chip_sim_cfg.hjson
Comment (msfschaffner) [reply hw/top_earlgrey/dv/chip_sim_cfg.hjson:257]: That's actually a good idea - let me look into this.

File: hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv
Comment (msfschaffner) [reply hw/top_earlgrey/dv/env/seq_lib/chip_stub_cpu_base_vseq.sv:52]: Ok that sounds good - will do.

**Comment by sriyerg** [comment]
> > If you dont mind, could you wait until my PR #15814 is merged?
> 
> It would be great if we could merge this fix so that the nightlies can pick this up... Do you expect yours to be merged soon?

I would like the nightlies to pick up mine as well (coverage boost!).  Plus I am starting to get tired of frequently resolving merge conflicts. :)

**Comment by msfschaffner** [comment]
> > > If you dont mind, could you wait until my PR #15814 is merged?
> > 
> > 
> > It would be great if we could merge this fix so that the nightlies can pick this up... Do you expect yours to be merged soon?
> 
> I would like the nightlies to pick up mine as well (coverage boost!). Plus I am starting to get tired of frequently resolving merge conflicts. :)

No worries, I know how annoying it is to keep large chip-level PRs up to date ;). And it looks like I need to do another pass on this PR anyways to address all feedback. Feel free to go ahead!

**Comment by msfschaffner** [comment]
Alright, I reworked this and got it finally running correctly. 
This creates and preloads a random ROM image with valid ECC and digest entirely in DV as part of the `dut_init()` task.

**Comment by msfschaffner** [comment]
@weicaiyang just reminded me that we already have a `update_rom_digest()` function, which makes this much cleaner. PTAL.

**Comment by msfschaffner** [comment]
The latest force push moves the rom init function to the `pre_start()` task.

**Review by weicaiyang — APPROVED** [review]
LGTM

File: hw/top_earlgrey/dv/env/seq_lib/chip_base_vseq.sv
Comment (sriyerg) [inline hw/top_earlgrey/dv/env/seq_lib/chip_base_vseq.sv:69]: Awesome, thanks @msfschaffner!

**Review by sriyerg — APPROVED** [review]
LGTM, thanks!

File: hw/top_earlgrey/dv/env/seq_lib/chip_base_vseq.sv
Comment (sriyerg) [inline hw/top_earlgrey/dv/env/seq_lib/chip_base_vseq.sv:203]: I will actually recommend moving this to `chip_stub_cpu_base_vseq` since this is overridden by the SW image in `chip_sw_base_vseq` anyway.

Comment (msfschaffner) [reply hw/top_earlgrey/dv/env/seq_lib/chip_base_vseq.sv:203]: But don't we want to randomize the ROM here with something? 
At least before we did a random init in `dut_init` in `chip_base_vseq`...

Comment (msfschaffner) [reply hw/top_earlgrey/dv/env/seq_lib/chip_base_vseq.sv:203]: If this needs tweaking, let's do that in a follow up (I have a PR that relies on this).

**Pull request merged**
